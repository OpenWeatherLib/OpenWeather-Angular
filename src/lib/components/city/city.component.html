<mat-card class="ga-card" *ngIf="city">
  <mat-card-header>
    <img mat-card-avatar class="ga-card-header-image" src="{{cityPictureUrl}}" alt="City pic">
    <mat-card-title>City</mat-card-title>
    <mat-card-subtitle>{{city?.name}}</mat-card-subtitle>
  </mat-card-header>
  <img mat-card-image src="{{cityPictureUrl}}" alt="City pic">
  <mat-card-content>
    <p>Country: {{city?.country}}</p>
    <p>Population: {{city?.population}}</p>
    <p>Coordinates: Lat: {{city?.coord?.lat}}, Lon: {{city?.coord?.lon}}</p>
  </mat-card-content>
</mat-card>
