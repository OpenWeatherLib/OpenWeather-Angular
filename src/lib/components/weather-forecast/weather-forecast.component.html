<span class="ga-display-inline-block">
  <mat-card class="ga-card-forecast-header" *ngIf="forecastWeather">
    <mat-card-header>
      <img mat-card-avatar class="ga-card-header-image" src="{{mostWeatherCondition?.icon}}" alt="Forecast weather pic">
      <mat-card-title>Forecast Weather</mat-card-title>
      <mat-card-subtitle>{{mostWeatherCondition?.description}}</mat-card-subtitle>
    </mat-card-header>
    <img mat-card-image src="{{mostWeatherCondition?.wallpaper}}" alt="Forecast weather pic">
  </mat-card>
</span>

<span class="ga-display-inline-block">
  <div class="ga-forecast-container" *ngIf="forecastWeather">
    <form class="ga-margin-1-rem">
      <mat-form-field class="ga-width-100-per">
        <input matInput placeholder="Search" [(ngModel)]="forecastWeatherSearch" [ngModelOptions]="{standalone: true}"
          (keyup)="searchForecastWeather()">
        <button mat-button matSuffix mat-icon-button aria-label="Clear" (click)="forecastWeatherSearch='';searchForecastWeather();">
          <mat-icon svgIcon="close"></mat-icon>
        </button>
      </mat-form-field>
    </form>

    <mat-list>
      <mat-list-item *ngFor="let forecastPart of forecastWeatherList" class="ga-card-forecast-weather-height-fix">
        <mat-card class="ga-card">
          <mat-card-header>
            <img mat-card-avatar class="ga-card-header-image" src="{{forecastPart?.weatherCondition?.icon}}" alt="Forecast weather pic">
            <mat-card-title>{{forecastPart?.dt_txt}}</mat-card-title>
            <mat-card-subtitle>{{forecastPart?.weatherCondition?.description}}</mat-card-subtitle>
          </mat-card-header>
          <mat-card-content>
            <p>Temperature: {{forecastPart?.main?.temp}}°C
              ({{forecastPart?.main?.temp_min}}°C-{{forecastPart?.main?.temp_max}}°C)</p>
            <p>Pressure: {{forecastPart?.main?.pressure}} mBar</p>
            <p>Humidity: {{forecastPart?.main?.humidity}} %</p>
          </mat-card-content>
        </mat-card>
      </mat-list-item>
    </mat-list>
  </div>
</span>
