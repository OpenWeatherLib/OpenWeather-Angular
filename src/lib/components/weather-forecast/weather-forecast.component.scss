@import '../../../styles/ga-common';

.ga-card-forecast-header {
  @extend .ga-card;
  margin-left: 2rem;
}

.ga-forecast-container {
  height: 90vh;
  overflow: overlay;
  width: 27rem;
}

// Default height for mat-list-item is 48px, but our cards have another height
.ga-card-forecast-weather-height-fix {
  height: 100% !important;
}
