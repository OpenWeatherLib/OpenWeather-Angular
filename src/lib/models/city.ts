import { GeoLocation } from "./geo-location";
import { classJsonKey } from "../decorator";

@classJsonKey("", "")
export class City {
  id: number = 0;
  name: string = "";
  country: string = "";
  population: number = 0;
  geoLocation: GeoLocation = new GeoLocation();

  isDefault(): boolean {
    return this.id === 0 && this.name === "" && this.country === "" && this.population === 0 && this.geoLocation.isDefault();
  }
}
