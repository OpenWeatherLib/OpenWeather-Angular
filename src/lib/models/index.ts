export * from "./address-component";
export * from "./city";
export * from "./city2";
export * from "./clouds";
export * from "./coordinates";
export * from "./coordinates2";
export * from "./geometry";
export * from "./main";
export * from "./rain";
export * from "./sys";
export * from "./weather-current";
export * from "./weather-forecast";
export * from "./weather-forecast-part";
export * from "./weather-part";
export * from "./wind";
export * from "./uv-index";
