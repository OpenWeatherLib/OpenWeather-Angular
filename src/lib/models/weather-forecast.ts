import { classJsonKey } from "../decorator";
import { City } from "./city";
import { WeatherForecastPart } from "./weather-forecast-part";

@classJsonKey(String().empty, String().empty)
export class WeatherForecast {
    city: City = new City();
    list: WeatherForecastPart[] = [];

    isDefault(): boolean {
        return this.city.isDefault() && this.list.length === 0;
    }
}
