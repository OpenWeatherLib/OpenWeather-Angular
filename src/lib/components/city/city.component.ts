import { Component, Input, OnInit } from "@angular/core";

import { BaseComponent } from "@lib/components/base-component/base.component";
import { City } from "@lib/models";
import { ImageService, OpenWeatherService } from "@lib/services";

@Component({
  selector: "ga-city",
  templateUrl: "./city.component.html",
  styleUrls: ["./city.component.scss"]
})
export class CityComponent extends BaseComponent implements OnInit {

  @Input() initialCityName: string = String().empty;

  city: City = null;
  cityPictureUrl: string = String().empty;

  constructor(private readonly imageService: ImageService,
    private readonly openWeatherService: OpenWeatherService) {
    super();
  }

  ngOnInit() {
    this.openWeatherService.loadCityData(this.initialCityName);

    this.registerSubscription(
      this.openWeatherService.city()
        .subscribe(city => {
          if (city) {
            this.city = city;
            this.imageService.receiveImagePictureUrl(this.city.name);
          }
        }));

    this.registerSubscription(
      this.imageService.cityPictureUrl()
        .subscribe(url => {
          this.cityPictureUrl = url;
        }));
  }
}
