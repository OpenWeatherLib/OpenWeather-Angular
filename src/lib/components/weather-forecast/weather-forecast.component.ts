import { Component, OnInit } from "@angular/core";
import { Store } from "@ngrx/store";
import { Observable } from "rxjs";

import WeatherCondition from "@lib/enums/weather-condition.enum";
import { WeatherForecastPart } from "@lib/models";
import { selectIsLoading, selectMostWeatherCondition, selectWeatherForecastList, setFilterRequestAction } from "@lib/store/weather-forecast-store";
import { RootState } from "@lib/store/root-state";

@Component({
  selector: "ga-weather-forecast",
  templateUrl: "./weather-forecast.component.html",
  styleUrls: ["./weather-forecast.component.scss"]
})
export class WeatherForecastComponent implements OnInit {

  isLoading$: Observable<boolean>;

  mostWeatherCondition$: Observable<WeatherCondition>;

  weatherForecastList$: Observable<WeatherForecastPart[]>;

  weatherForecastSearch: string = "";

  readonly filterWeatherForecast = (): void => this.store$.dispatch(setFilterRequestAction({ filter: this.weatherForecastSearch }));

  readonly trackByIndex = (index: number, _: WeatherForecastPart) => index;

  constructor(private readonly store$: Store<RootState>) { }

  ngOnInit() {
    this.isLoading$ = this.store$.select(selectIsLoading);

    this.mostWeatherCondition$ = this.store$.select(selectMostWeatherCondition);

    this.weatherForecastList$ = this.store$.select(selectWeatherForecastList);
  }
}
