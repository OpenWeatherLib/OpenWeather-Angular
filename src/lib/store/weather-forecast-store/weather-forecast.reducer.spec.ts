import WeatherCondition from "@lib/enums/weather-condition.enum";
import { City, WeatherForecast } from "@lib/models";
import { loadCitySuccessAction } from "../city-store/city.actions";
import { loadWeatherForecastRequestAction, loadWeatherForecastSuccessAction, loadWeatherForecastErrorAction, setFilterRequestAction, clearFilterRequestAction } from "./weather-forecast.actions";
import { weatherForecastReducer } from "./weather-forecast.reducer";
import { WeatherForecastState } from "./weather-forecast.state";

describe("WeatherForecast Reducer Tests", () => {

    const initialState: WeatherForecastState = {
        weatherForecast: undefined,
        weatherForecastList: [],
        mostWeatherCondition: WeatherCondition.null,
        filter: undefined,
        isLoading: false,
        error: "Test Error"
    };

    describe("State Changes", () => {
        test("should have initial state after invalid action", () => {
            // Arrange & Act
            const testState: WeatherForecastState = weatherForecastReducer(initialState, { type: "INVALID_ACTION" } as any);

            // Assert
            expect(testState).toBe(initialState);
        });

        test("should load city on success of loading", () => {
            // Arrange
            const loadRequest = loadWeatherForecastRequestAction({ city: {} as City });
            const loadSuccess = loadWeatherForecastSuccessAction({ weatherForecast: {} as WeatherForecast });

            // Act
            const testState: WeatherForecastState = [loadRequest, loadSuccess].reduce(weatherForecastReducer, initialState);

            // Assert
            expect(testState).toMatchSnapshot();
        });

        test("should set error on error of loading", () => {
            // Arrange
            const loadRequest = loadWeatherForecastRequestAction({ city: {} as City });
            const loadError = loadWeatherForecastErrorAction({ error: "Error" });

            // Act
            const testState: WeatherForecastState = [loadRequest, loadError].reduce(weatherForecastReducer, initialState);

            // Assert
            expect(testState).toMatchSnapshot();
        });

        test("should set loading to true on loadCitySuccessAction", () => {
            // Arrange
            const loadSuccess = loadCitySuccessAction({ city: {} as City });

            // Act
            const testState: WeatherForecastState = [loadSuccess].reduce(weatherForecastReducer, initialState);

            // Assert
            expect(testState).toMatchSnapshot();
        });

        test("should update state on setFilterRequestAction", () => {
            // Arrange
            const filterAction = setFilterRequestAction({ filter: "Filter" });

            // Act
            const testState: WeatherForecastState = [filterAction].reduce(weatherForecastReducer, initialState);

            // Assert
            expect(testState).toMatchSnapshot();
        });

        test("should update state on clearFilterRequestAction", () => {
            // Arrange
            const clearFilterAction = clearFilterRequestAction();

            // Act
            const testState: WeatherForecastState = [clearFilterAction].reduce(weatherForecastReducer, initialState);

            // Assert
            expect(testState).toMatchSnapshot();
        });
    });
});
