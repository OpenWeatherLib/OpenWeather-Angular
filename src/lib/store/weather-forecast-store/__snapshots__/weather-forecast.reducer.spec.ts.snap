// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherForecast Reducer Tests State Changes should load city on success of loading 1`] = `
Object {
  "error": undefined,
  "filter": undefined,
  "isLoading": false,
  "mostWeatherCondition": Object {
    "count": 0,
    "description": "Null",
    "icon": "/assets/weather_dummy.png",
    "id": 0,
    "wallpaper": "/assets/weather_wallpaper_dummy.png",
  },
  "weatherForecast": Object {},
  "weatherForecastList": undefined,
}
`;

exports[`WeatherForecast Reducer Tests State Changes should set error on error of loading 1`] = `
Object {
  "city": undefined,
  "error": "Error",
  "filter": undefined,
  "isLoading": false,
  "mostWeatherCondition": Object {
    "count": 0,
    "description": "Null",
    "icon": "/assets/weather_dummy.png",
    "id": 0,
    "wallpaper": "/assets/weather_wallpaper_dummy.png",
  },
  "weatherForecast": undefined,
  "weatherForecastList": Array [],
}
`;

exports[`WeatherForecast Reducer Tests State Changes should set loading to true on loadCitySuccessAction 1`] = `
Object {
  "error": "Test Error",
  "filter": undefined,
  "isLoading": true,
  "mostWeatherCondition": Object {
    "count": 0,
    "description": "Null",
    "icon": "/assets/weather_dummy.png",
    "id": 0,
    "wallpaper": "/assets/weather_wallpaper_dummy.png",
  },
  "weatherForecast": undefined,
  "weatherForecastList": Array [],
}
`;

exports[`WeatherForecast Reducer Tests State Changes should update state on clearFilterRequestAction 1`] = `
Object {
  "error": "Test Error",
  "filter": undefined,
  "isLoading": false,
  "mostWeatherCondition": Object {
    "count": 0,
    "description": "Null",
    "icon": "/assets/weather_dummy.png",
    "id": 0,
    "wallpaper": "/assets/weather_wallpaper_dummy.png",
  },
  "weatherForecast": undefined,
  "weatherForecastList": Array [],
}
`;

exports[`WeatherForecast Reducer Tests State Changes should update state on setFilterRequestAction 1`] = `
Object {
  "error": "Test Error",
  "filter": "Filter",
  "isLoading": false,
  "mostWeatherCondition": Object {
    "count": 0,
    "description": "Null",
    "icon": "/assets/weather_dummy.png",
    "id": 0,
    "wallpaper": "/assets/weather_wallpaper_dummy.png",
  },
  "weatherForecast": undefined,
  "weatherForecastList": Array [],
}
`;
